src/python/WMCore/ReqMgr/CherryPyThreads/BuildParentLock.py

```diff
@@ -0,0 +1,102 @@
+find which requests require parent dataset to be processed;
+and run the DBS parentage lookup.
+"""
+from __future__ import division, print_function
+
+from WMCore.REST.CherryPyPeriodicTask import CherryPyPeriodicTask
+from WMCore.Services.DBS.DBSReader import DBS3Reader
+from WMCore.Services.RequestDB.RequestDBReader import RequestDBReader
+from WMCore.Services.ReqMgrAux.ReqMgrAux import ReqMgrAux
+
+
+class BuildParentLock(CherryPyPeriodicTask):
+
+    def __init__(self, rest, config):
+
+        super(BuildParentLock, self).__init__(config)
+        self.reqmgrAux = ReqMgrAux(config.reqmgr2_url, logger=self.logger)
+        self.dbs = DBS3Reader(config.dbs_url)
+        self.datasetToParentMap = {}
+
+
+    def setConcurrentTasks(self, config):
+        """
+        sets the list of functions which
+        """
+        self.concurrentTasks = [{'func': self.fetchIncludeParentsRequests,
+                                 'duration': config.updateParentsInterval}]
+
+    def fetchIncludeParentsRequests(self, config):
+        """
+        Fetch active requests from the "requestsincludeparents" couch view that
+        have IncludeParents=True, find parents of each dataset and send to
+        reqmgr2 auxiliary database.
+        """
+        # use this boolean to signal whether there were datasets that failed
+        # to get their parentage resolved
+        incompleteParentage = False
+        # use this boolean to signal if new parent datasets need to be locked
+        auxDbUpdateRequired = False
+
+        setDsets = set()
+        setParents = set()
+        dictParents = {}
+
+        self.logger.info("Executing parent lock cherrypy thread")
+
+        # query couch view to find datasets for workflows requiring parent datasets
+        # only returning requests with the statuses in filterKeys
+        reqDB = RequestDBReader(config.reqmgrdb_url)
+        filterKeys = ['assignment-approved', 'assigned', 'staging', 'staged', 'acquired',
+                      'running-open', 'running-closed', 'force-complete', 'completed']
+        results = reqDB._getCouchView("requestsincludeparents", {}, filterKeys)
+        for row in results["rows"]:
+            dataset = row["value"]
+            setDsets.add(dataset)
+
+        # check to see if any changes have been made
+        # this will be true where new datasets are added and old datasets are no longer needed
+        if setDsets ^ set(self.datasetToParentMap.keys()):
+            auxDbUpdateRequired = True
+
+        self.logger.info("Found %d unique datasets requiring the parent dataset", len(setDsets))
+        if auxDbUpdateRequired:
+            self.logger.info("Found new parent dataset locks to update.")
+            # reset dataset to parent dataset map
+            self.datasetToParentMap = {}
+            # look up parent datasets via DBS
+            for dset in setDsets:
+                try:
+                    res = self.dbs.listDatasetParents(dset)
```

Comment:
This is actually the call I planned to have cached in memory. Parent dataset will not change, so we resolve it once and use it from the cache while the process exists.
Which also means L68 (resetting the map variable) should be removed, such that we "never" cleanup that structure. Memory footprint should be negligible.